go/pools/rpc_pool.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pools
+
+import (
+	"context"
+	"time"
+)
+
+// RPCPool is an abstraction on top of ResourcePool to throttle concurrent
+// operations (RPCs in the most common case) to prevent excessive fanout.
+type RPCPool interface {
```

Comment:
I'm pondering the value that this extra interface is adding. It seems that we could go use the resource pool directly and the overall code doesn't change significantly. By using the resource pool directly, I mean on this context: 
```	
cluster.GetSchemaPool = pools.NewRPCPool(getSchemaRPCPoolSize, getSchemaRPCPoolWaitTimeout)
```

If we can avoid the boiler plate, that seems a better option. Later, if this becomes a pattern that is used multiple times, we can add this abstraction and refactor.

Reply:
So, this is going to become a pattern that's used multiple times, because in the very next change I have queued up in this line of work is to introduce a second pool for topo RPCs, and use that _everywhere_ in vtadmin. The main reason I didn't do it all in one shot was to keep the diffs manageable, so the order of operations here is:
1. introduce the pattern and use it in one, well-contained place
2. use the pattern in the rest of the places

The other thing about "we should use the pool directly" is that there's no way to provide an "unbounded" behavior, which this implementation currently does. If we remove this abstraction, then this code has to become actually:

```go
func (c *Cluster) GetSchema(ctx context.Context, ...) ... {
    ....

    var f func() (*vtctldatapb.GetSchemaResponse, error)

    if c.GetSchemaPool == nil {
        f = func() (*vtctldatapb.GetSchemaResponse, error) {
            return c.Vtctld.GetSchema(ctx, ...)
        }
    } else {
        f = func() (*vtctldatapb.GetSchemaResponse, error) {
            rp, err := c.GetSchemaPool.Get(ctx)
            if err != nil {
                return nil, err
            }
            
            defer c.GetSchemaPool.Put(rp)
            return c.Vtctld.GetSchema(ctx, ...)
        }
    }
    
    resp, err = f()
}
```

and then every single place we want to potentially use a pool has to follow this if/else pattern, vs being able to hide the unbounded case behind the interface

Reply:
I see. Thank you for all the context. The reasoning makes sense and I agree that without this, it adds complexity for the unbounded case. However, the example assumes there is a need for an unbounded pool. Why do we need that? If we are going to introduce the concept of pool, if you don't need one you could effectively achieve the same by providing a very large size to the pool. 

It is still not clear how important is this abstraction. Too many abstractions if something that we have struggled in the Vitess codebase and on this case having unbounded functionality doesn't seem compelling enough.